src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -300,59 +269,67 @@ class CassandraJournal(cfg: Config) extends AsyncWriteJournal with CassandraReco
-    val toSeqNr = math.min(toSequenceNr, readHighestSequenceNr(persistenceId, fromSequenceNr))
-    val highestPartition = partitionNr(toSeqNr) + 1 // may have been moved to the next partition
-    val partitionInfos = (lowestPartition to highestPartition).map(partitionInfo(persistenceId, _, toSeqNr))
-
-    val logicalDelete = session.executeAsync(
-      cassandraSession.preparedInsertDeletedTo.bind(persistenceId, toSeqNr: JLong)
-    )
-
-    if (config.cassandra2xCompat) {
-      def asyncDeleteMessages(partitionNr: Long, messageIds: Seq[MessageId]): Future[Unit] = executeBatch({ batch =>
-        messageIds.foreach { mid =>
-          batch.add(cassandraSession.preparedDeleteMessages
-            .bind(mid.persistenceId, partitionNr: JLong, mid.sequenceNr: JLong))
-        }
-      }, deleteRetryPolicy)
-
-      partitionInfos.map(future => future.flatMap(pi => {
-        Future.sequence((pi.minSequenceNr to pi.maxSequenceNr).grouped(config.maxMessageBatchSize).map { group =>
-          {
-            val delete = asyncDeleteMessages(pi.partitionNr, group map (MessageId(persistenceId, _)))
-            delete.onFailure {
-              case e => log.warning(s"Unable to complete deletes for persistence id ${persistenceId}, toSequenceNr ${toSequenceNr}. The plugin will continue to function correctly but you will need to manually delete the old messages.", e)
+    asyncHighestDeletedSequenceNumber(persistenceId).flatMap { highestDeletedSequenceNumber =>
+      asyncReadLowestSequenceNr(persistenceId, 1L, highestDeletedSequenceNumber).flatMap { lowestSequenceNr =>
+        super.asyncReadHighestSequenceNr(persistenceId, lowestSequenceNr).flatMap { highestSequenceNr =>
+          val lowestPartition = partitionNr(lowestSequenceNr)
+          val toSeqNr = math.min(toSequenceNr, highestSequenceNr)
+          val highestPartition = partitionNr(toSeqNr) + 1 // may have been moved to the next partition
+
+          val boundInsertDeletedTo = preparedInsertDeletedTo.map(_.bind(persistenceId, toSeqNr: JLong))
+          val logicalDelete = boundInsertDeletedTo.flatMap(session.executeWrite)
+
+          def partitionInfos = (lowestPartition to highestPartition).map(partitionInfo(persistenceId, _, toSeqNr))
+
+          def physicalDelete(): Unit = {
+            if (config.cassandra2xCompat) {
+              def asyncDeleteMessages(partitionNr: Long, messageIds: Seq[MessageId]): Future[Unit] = {
+                val boundStatements = messageIds.map(mid =>
+                  preparedDeleteMessages.map(_.bind(mid.persistenceId, partitionNr: JLong, mid.sequenceNr: JLong)))
+                Future.sequence(boundStatements).flatMap { stmts =>
+                  executeBatch(batch => stmts.foreach(batch.add), deleteRetryPolicy)
+                }
+              }
+
+              partitionInfos.map(future => future.flatMap(pi => {
+                Future.sequence((pi.minSequenceNr to pi.maxSequenceNr).grouped(config.maxMessageBatchSize).map { group =>
```

Comment:
@ktoso looks like this is wrong. `pi.maxSequenceNr` should be replaced with `min(toSeqNr, pi.maxSequenceNr`), I think